Music: Redesign Part 2

tituscmd opened the pull request:
Hey everybody,

Since creating PR #2292 and introducing a small little redesign of the music app, I've had some more time on my hands and finally came around to creating a better version of it.

I've moved the artist and track name up a bit to add a new progress bar that shows the progress of the song you're listening to, with the time that has passed shown on the left and the time remaining shown on the right.
Sadly, I've had to remove (comment out) the disc animation in the top right because of space issues. If anyone is especially fond of the disc I'm sure space can be made to fit it back in. It is definitely not a final decision by me, hence why it's only commented out and not removed altogether.
But now that I've temporarily removed the disc animation, the top of the screen seemed quite empty to me. And after some thinking and digging through the code, I stumbled upon these lines of code in the Music.cpp:
```
294 - // Let's assume it starts playing instantly
295 - // TODO: In the future should check for BT connection for better UX
```
And so I thought "why not do that?" and I added an indicator text at the top telling you if you are connected via bluetooth. Please share your opinion on this.


Finally, here are some pictures of it all:
![WhatsApp Image 2025-12-28 at 01 30 44](https://github.com/user-attachments/assets/87327a5a-8c42-4310-9de3-8967dc5479f3)
<img width="288" height="288" alt="Screenshot 2025-12-28 at 01-34-08 InfiniEmu" src="https://github.com/user-attachments/assets/3b854126-ea90-4dd8-861b-da51143c5863" />

tituscmd committed:
Merge pull request #26 from InfiniTimeOrg/main

Alarm: Simplify alarm alerting screen (#2211)

tituscmd committed:
music app redesign

tituscmd committed:
Merge branch 'InfiniTimeOrg:main' into music_redesign

tituscmd committed:
removed an empty line

tituscmd committed:
Merge branch 'music_redesign' of github.com:tituscmd/InfiniTime into music_redesign

tituscmd committed:
Some tweaks for the default texts, as well as adding a different approach to switching between skipping and volume buttons

tituscmd committed:
formatting

tituscmd committed:
more formatting

tituscmd committed:
Merge branch 'main' into music_redesign

tituscmd committed:
revert back to swiping as a gesture to switch between skip and volume buttons

tituscmd committed:
remove all unneccessary references to the motor controller

tituscmd committed:
formatting

tituscmd committed:
Merge branch 'main' into music_redesign

tituscmd committed:
removed accidental change

tituscmd committed:
Merge branch 'music_redesign' of github.com:tituscmd/InfiniTime into music_redesign

tituscmd committed:
Merge branch 'main' into music_redesign

tituscmd committed:
Merge branch 'main' into music_redesign

tituscmd committed:
correct color of the buttons

tituscmd committed:
remove redundant initialisers for trackName and albumName

tituscmd committed:
Merge branch 'main' into music_redesign

tituscmd committed:
Merge branch 'main' into music_redesign

tituscmd committed:
Merge branch 'main' into music_redesign

tituscmd committed:
add progress bar

tituscmd committed:
Update settings.json

tituscmd committed:
Merge branch 'main' into music_redesign_2

tituscmd committed:
formatting

tituscmd committed:
Merge branch 'music_redesign_2' of github.com:tituscmd/InfiniTime into music_redesign_2

tituscmd committed:
formatting

tituscmd committed:
more formatting

liamcharger commented:
Definitely like this better than the current UI!

But what if you replaced the "connected" with the status bar/icons instead? That makes more sense imo

P.S I always thought the disc was a waste of space :)

mark9064 commented:
Aw I quite like the little disc

Looks like a good change overall

I interpret the comment about Bluetooth as clearing the presented data when Bluetooth is disconnected, rather than showing an indicator. Since if Bluetooth isn't connected, we don't know anything about the music state and none of the controls are going to do anything. So I think it might be better UX to disable the controls and show "No connection" when Bluetooth is down rather than having a status bar, that way you get that space back. What do you think?

tituscmd commented:
I think that's a good idea.

The initial reason for that status bar/"Connected" text was the fact that the little disc physically didn't fit with the new design, but without it the top seemed so empty.

I'll share some screenshots of the different ideas talked about in this PR later when I have the time again. Then we can see and decide which one we like best.
How does that sound? 🙂

mark9064 commented:
Sounds good to me :)

tituscmd committed:
different design for music.cpp

tituscmd commented:
Im not necessarily against any of those ideas, it just seems to me that, from a UI perspective, the status bar is only ever present in watchfaces or "transition screens" i.e. quick settings, app launcher. 
I feel like an app should feel like its own separate entity from those, so including the status bar breaks that.

I like the idea of moving the time to the left and adding the disc back, but it's ever so slightly too large for the new UI. Would it be easy/doable to make that disc a bit smaller?

mark9064 commented:
Yeah more broadly there is #1592

The disc could be resized, it's a small PNG that's been converted to LVGL format. The original PNG is in the tree (icons/music/disc.png)

tituscmd commented:
> I agree with @mark9064 here about the status bar. It shouldn't be exclusive to the launcher and quick settings (usually there just isn't enough space). I can think of too many times I would've liked to check the time while I was inside an app I didn't want to close

I agree partly. In an ideal world, the status bar should be visible in any app. But since that is not the case right now, I think it’s better to stick to this pattern until we find a better solution for every app/globally, so everything remains predictable for now.

tituscmd committed:
renaming to camelCase, time top center, some visual tweaks

tituscmd committed:
tweak code for much easier alignment testing

tituscmd committed:
tweak code for much easier alignment testing

tituscmd committed:
change controls button

tituscmd committed:
tweaks to the new swap controls button

LinuxinaBit commented:
What about just this instead?

<img width="288" height="288" alt="image" src="https://github.com/user-attachments/assets/3bcee6d3-bfab-4f7f-947f-dc9caccdd8c7" />

No need to make one tap into two, I figure...

tituscmd commented:
> > ![image](https://github.com/user-attachments/assets/937d3ff6-e8c4-4e25-9bb4-ec8d4e48fa54)
> > Looks like this on-device, I quite like it
> 
> While it informs the user of the volume controls, I don't think that's intuitive either. The hint itself looks like a button, and I don't like the idea of it being permanent. 
It is indeed a button and achieves the same as the "old" swiping gesture, but as a button

> @LinuxinaBit 's solution is great, although the button colors should be uniform. We don't lose any current elements, and it fixes the entire issue. 
Very fair.
How about we place the buttons below the current three buttons though, so it looks more like it belongs to those control buttons?

tituscmd commented:
>How about we place the buttons below the current three buttons though, so it looks more like it belongs to those control buttons?

I like this a lot actually. I've changed the height of the buttons to match the InfiniTime Vision on UI and I think it looks really good.
<img width="288" height="288" alt="image" src="https://github.com/user-attachments/assets/e0ca69e8-9a08-4fea-984b-a6f5a674fbaf"/>

tituscmd committed:
Different version of the music app, where all buttons are on one screen

LinuxinaBit commented:
The reason I didn’t group the buttons like that is so the touch targets could be made bigger (expanding down past the visual button), and the extra space provides some good visual separation, both sorely needed on a screen this small.

As for the colors, I themed it the same as your `Volume Controls` button to provide some visual balance and avoid making it too uniform (the volume icons should’ve actually been a bit more gray to match, but I forgot to do that).
I think it looked a lot better than when I tried making them the same color/shade.

Main thing is that I really wanted to avoid making it just another uniform grid of icons because it just doesn’t look all that good, and is less usable to boot.

Oh, also it move the song title/artist text a bit too high for my preference...

tituscmd commented:
> As for the colors, I themed it the same as your `Volume Controls` button to provide some visual balance and avoid making it too uniform (the volume icons should’ve actually been a bit more gray to match, but I forgot to do that). I think it looked a lot better than when I tried making them the same shade.

The darker color is actually already used when the watch is disconnected, showing that they won't work. As far as I can tell, that's been the visual cue for the user to understand that the button currently won't work (see stopwatch app -> lap button)


> Main thing is that I really wanted to avoid making it just another grid of icons because it just doesn’t look all that good, and is less usable to boot.

Not sure what you mean by less usable to boot. Otherwise, the grid of icons is down to preference I think. I think having the volume buttons up at the top separates them too much from the controls and makes them look out of place.

LinuxinaBit commented:
I think part of the problem is that the spacing isn’t consistent with the bottom ones, and they aren’t tall enough. I think fixing both of those issues will help.

LinuxinaBit commented:
It might also be worth just making them a *different* shade (not the disabled button or the standard button shade) just for visual variety.
I think it could end up looking really good if you find the right combination.

tituscmd commented:
With double the spacing between the two sets of buttons, it looks a lot better imo.
Now the only downside is that the rest is really smushed into the top edge of the screen

![image](https://github.com/user-attachments/assets/5765fc97-bb2a-4d54-a4af-751f6929701e)

tituscmd committed:
another UI iteration, my favorite (volume buttons above track buttons)

tituscmd commented:
Another pic from on my wrist. I think it looks good!
![image](https://github.com/user-attachments/assets/b68500b9-d85c-4747-af58-85ebc66b26e2)

LinuxinaBit commented:
Those are so old they don't even show the taller height buttons used on the stopwatch etc.
I think when combining buttons in the same UI, consistency within the UI is far more important anyway.


On August 22, 2025 6:18:08 PM CDT, Titus ***@***.***> wrote:
>tituscmd left a comment (InfiniTimeOrg/InfiniTime#2337)
>
><img width="990" height="240" alt="example" src="https://github.com/user-attachments/assets/3fba4dbe-6039-4de9-bfbe-06ca7314a86e" />
>
>These are the UI guidelines I was going off of. That's why the gap is bigger. The pixels on the images I uploaded aren't exactly the same because, for whatever reason, they were rendered as 288x288 instead of 240x240.
>
>-- 
>Reply to this email directly or view it on GitHub:
>https://github.com/InfiniTimeOrg/InfiniTime/pull/2337#issuecomment-3215929834
>You are receiving this because you were mentioned.
>
>Message ID: ***@***.***>

tituscmd committed:
version with buttons up top

dariusarnold commented on src/displayapp/screens/Music.cpp:
```diff
@@ -24,6 +24,7 @@
 #include "displayapp/icons/music/disc_f_1.c"
```
These disc files can probably removed when fully unused.

tituscmd replied:
Yep! Good catch

dariusarnold commented:
@tituscmd Can I ask about the Track/Artist info (currently placeholder "Some Track")? Is the "Disconnected" replaced with the artist when connected? I liked the two line design from the first post, since it would reduce text scrolling.

---

The current commit ed30e5 from https://github.com/InfiniBros/InfiniTime/tree/music_redesign_2 makes InfiniSim crash on my end when entering the music app. It happens reliably with the following callstack:
```
draw_indic(lv_obj_t * bar, const lv_area_t * clip_area) (/workspaces/InfiniTime/src/libs/lvgl/src/lv_widgets/lv_bar.c:498)
lv_bar_design(lv_obj_t * bar, const lv_area_t * clip_area, lv_design_mode_t mode) (/workspaces/InfiniTime/src/libs/lvgl/src/lv_widgets/lv_bar.c:378)
lv_refr_obj(lv_obj_t * obj, const lv_area_t * mask_ori_p) (/workspaces/InfiniTime/src/libs/lvgl/src/lv_core/lv_refr.c:895)
lv_refr_obj(lv_obj_t * obj, const lv_area_t * mask_ori_p) (/workspaces/InfiniTime/src/libs/lvgl/src/lv_core/lv_refr.c:935)
lv_refr_obj_and_children(lv_obj_t * top_p, const lv_area_t * mask_p) (/workspaces/InfiniTime/src/libs/lvgl/src/lv_core/lv_refr.c:837)
lv_refr_area_part(const lv_area_t * area_p) (/workspaces/InfiniTime/src/libs/lvgl/src/lv_core/lv_refr.c:768)
lv_refr_area(const lv_area_t * area_p) (/workspaces/InfiniTime/src/libs/lvgl/src/lv_core/lv_refr.c:510)
lv_refr_areas() (/workspaces/InfiniTime/src/libs/lvgl/src/lv_core/lv_refr.c:400)
_lv_disp_refr_task(lv_task_t * task) (/workspaces/InfiniTime/src/libs/lvgl/src/lv_core/lv_refr.c:199)
lv_task_exec(lv_task_t * task) (/workspaces/InfiniTime/src/libs/lvgl/src/lv_misc/lv_task.c:386)
lv_task_handler() (/workspaces/InfiniTime/src/libs/lvgl/src/lv_misc/lv_task.c:134)
Pinetime::Applications::DisplayApp::Refresh(class Pinetime::Applications::DisplayApp * const this) (/workspaces/InfiniTime/src/displayapp/DisplayApp.cpp:260)
Pinetime::Applications::DisplayApp::Process(void * instance) (/workspaces/InfiniTime/src/displayapp/DisplayApp.cpp:149)
sdl_function_wrapper(void * instance) (/workspaces/InfiniSim/sim/task.cpp:61)
libSDL2-2.0.so.0![Unknown/Just-In-Time compiled code] (Unknown Source:0)
libc.so.6!start_thread(void * arg) (pthread_create.c:447)
libc.so.6!clone() (clone.S:100)
```

The debugger (gdb) prints
```
Thread 37 "displayapp" received signal SIGFPE, Arithmetic exception.
[Switching to Thread 0x7fb1d7fff6c0 (LWP 5533)]
0x00005625e98d981c in draw_indic (bar=0x7fb1cc001d58, clip_area=0x7fb1d7ffe8f8) at /workspaces/InfiniTime/src/libs/lvgl/src/lv_widgets/lv_bar.c:498
498	        anim_start_value_x = (int32_t)((int32_t)anim_length * (ext->start_value - ext->min_value)) / range;
```

I can see that the issue is caused by `range` being zero when attached in gdb. I am not familiar enough with InfiniSim + LVGL to diagnose this quickly though. It might be a simulator-only issue as well, I have been using an earlier commit of this on the watch for some time now where it worked well.

dariusarnold commented:
@tituscmd Great, looking forward to try it on the watch.

tituscmd committed:
clean up code

tituscmd committed:
finally got a formatting extension, so this is hopefully the last formatting commit

tituscmd committed:
finally got a formatting extension, so this is hopefully the last formatting commit (ironically one more)

tituscmd committed:
some final spacing optimizations

tituscmd committed:
remove some more (now) unused lines

tituscmd committed:
lots of code optimization, mainly to get the reconnection smoother

tituscmd commented:
![image](https://github.com/user-attachments/assets/dfcbd348-290c-4a1f-96b7-38f85a650cc3)
![image](https://github.com/user-attachments/assets/55b5d055-d457-4da9-9522-ab8e9edafb22)
![image](https://github.com/user-attachments/assets/49d5f080-4376-4103-9085-cddf022487af)

tituscmd commented:
![WhatsAppVideo2025-08-25at23 43 48-ezgif com-video-to-gif-converter(1)](https://github.com/user-attachments/assets/5d242f17-0d6d-4128-9c5e-9c1a451aa0d6)

liamcharger commented:
Looking good, @tituscmd!

Thanks for your work on this!

tituscmd commented:
This PR should be ready for review from my side :)

LinuxinaBit commented:
Ooh, what if the progress bar was blue?

<img width="288" height="288" alt="image" src="https://github.com/user-attachments/assets/b6306cba-8a26-40c4-8c07-f548655c74d7" />

tituscmd committed:
remove unused emoji

tituscmd committed:
Merge branch 'main' into music_redesign_2

LinuxinaBit commented:
Yeah, not a big deal. Though it isn't really for readability; it's just about adding some visual flair to an otherwise colorless interface :)

On September 13, 2025 6:51:38 AM CDT, Titus ***@***.***> wrote:
>tituscmd left a comment (InfiniTimeOrg/InfiniTime#2337)
>
>> Ooh, what if the progress bar was blue?
>> <img alt="image" width="288" height="288" src="https://private-user-images.githubusercontent.com/105567407/486444645-b6306cba-8a26-40c4-8c07-f548655c74d7.png?jwt=eyJ0eXAiOiJKV1QiLCJhbGciOiJIUzI1NiJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3NTc3NjQxNTQsIm5iZiI6MTc1Nzc2Mzg1NCwicGF0aCI6Ii8xMDU1Njc0MDcvNDg2NDQ0NjQ1LWI2MzA2Y2JhLThhMjYtNDBjNC04YzA3LWY1NDg2NTVjNzRkNy5wbmc_WC1BbXotQWxnb3JpdGhtPUFXUzQtSE1BQy1TSEEyNTYmWC1BbXotQ3JlZGVudGlhbD1BS0lBVkNPRFlMU0E1M1BRSzRaQSUyRjIwMjUwOTEzJTJGdXMtZWFzdC0xJTJGczMlMkZhd3M0X3JlcXVlc3QmWC1BbXotRGF0ZT0yMDI1MDkxM1QxMTQ0MTRaJlgtQW16LUV4cGlyZXM9MzAwJlgtQW16LVNpZ25hdHVyZT01OWIxZWJmNDkxMDVhYzNmN2FjMDM4N2Y5NTcxMjY1MTRmZDExODY4YzdjOGRkMGYyZWQwMmI5MGQ2MjcyODlhJlgtQW16LVNpZ25lZEhlYWRlcnM9aG9zdCJ9.lfZgOm-T2nHsqv1aP-lkOgWbtc8dktR47Pqy8nZbCFM">
>
>I personally like the gray and don't see any big improvement in terms of readability with a blue one.
>
>-- 
>Reply to this email directly or view it on GitHub:
>https://github.com/InfiniTimeOrg/InfiniTime/pull/2337#issuecomment-3288150660
>You are receiving this because you were mentioned.
>
>Message ID: ***@***.***>

tituscmd committed:
Merge branch 'main' into music_redesign_2

mark9064 commented on src/displayapp/screens/Music.cpp:
```diff
@@ -150,74 +158,94 @@ Music::~Music() {
 
 void Music::UpdateLength() {
+  int remaining = totalLength - currentPosition;
+  if (remaining < 0)
```
Could you add braces here

mark9064 replied:
On the if statements:
```cpp
if (remaining < 0) {
...
}
```

mark9064 commented on src/displayapp/screens/Music.cpp:
```diff
@@ -150,74 +158,94 @@ Music::~Music() {
+  bleState = bleController.IsConnected();
+  if (bleState.Get() == false) {
+    SetDisconnectedUI();
+    lastConnected = false;
```
Could `lastConnected` be avoided by relying on `bleState.IsUpdated()`?

mark9064 commented on src/displayapp/screens/Music.cpp:
```diff
@@ -150,74 +158,94 @@ Music::~Music() {
+  }
+}
+
+void Music::SetDisconnectedUI() {
```
If the state variables (playing, artist, track, album etc) were updated here, could we get rid of force? I think the UI will be more robust if the state variables are always kept up to date, that way connection/disconnection don't need special handling

Would also need handling during in the constructor

mark9064 commented on src/displayapp/screens/Symbols.h:
```diff
@@ -42,6 +42,7 @@ namespace Pinetime {
         static constexpr const char* sleep = "\xEE\xBD\x84";
         static constexpr const char* calculator = "\xEF\x87\xAC";
         static constexpr const char* backspace = "\xEF\x95\x9A";
+        static constexpr const char* swap = "\xef\x81\xb9";
```
Is this used anywhere?

mark9064 reviewed (commented):
Looking good :)

Haven't tested locally or ran clang-tidy yet

tituscmd committed:
Update Music.cpp

Co-authored-by: mark9064 <30447455+mark9064@users.noreply.github.com>

tituscmd committed:
react to review, most notably change artist, track and album to use DirtyValue

tituscmd committed:
Merge branch 'music_redesign_2' of github.com:InfiniBros/InfiniTime into music_redesign_2

tituscmd committed:
Merge branch 'main' into music_redesign_2

tituscmd committed:
use bleState.isUpdated() instead of manual lastConnected approach

tituscmd committed:
Merge branch 'music_redesign_2' of github.com:InfiniBros/InfiniTime into music_redesign_2

tituscmd committed:
change behavior of Refresh cycle

tituscmd committed:
I'll never understand why it does this

mark9064 commented on src/displayapp/screens/Music.cpp:
```diff
@@ -20,10 +20,8 @@
 #include "components/ble/MusicService.h"
-#include "displayapp/icons/music/disc.c"
-#include "displayapp/icons/music/disc_f_1.c"
-#include "displayapp/icons/music/disc_f_2.c"
```
The disc icon should be deleted as well

tituscmd replied:
I'm also not particularly fond of getting rid of it, but it makes too much sense for usability 😔

tituscmd replied:
Maybe one day we will find a way to implement them, it feels wrong to get rid of them entirely...

mark9064 replied:
I think our code policy is generally if it isn't used, it should be removed. The Git history will remember, we can always bring it back. Mainly as it's confusing to keep dead code around: someone browsing the source after this change might see these files and then try to figure out where they're from.

If there's ever a chance to use them again, being completely realistic I doubt they're going to be remembered any more than they would be by continuing to exist in tree. And by then they'll also probably be incompatible with a new LVGL version or whatever. In short keeping unused code around is confusing and causes problems later as it becomes incompatible with the evolving codebase

tituscmd replied:
Sounds reasonable. I've removed the files in the latest commit and believe this PR should be done now.

mark9064 commented on src/displayapp/screens/Music.cpp:
```diff
@@ -152,22 +156,71 @@ Music::~Music() {
+  }
+
+  album = musicService.getAlbum();
+  if (album.IsUpdated()) {
```
Album unused? If so we can remove it as a member

mark9064 replied:
I don't think it's too important compared to the track title and artist so it probably doesn't justify the space. Especially as a lot of the time the field is probably blank (since podcasts/video content etc. don't really have an album)

tituscmd replied:
Yep, that's fair. I'll go ahead and remove it next time I'm available. Might take a few days because of the upcoming holidays.

mark9064 replied:
No rush, drop a message whenever :) happy holidays!

mark9064 commented on src/displayapp/screens/Music.h:
```diff
@@ -60,30 +68,29 @@ namespace Pinetime {
 
-        Widgets::PageIndicator pageIndicator = Widgets::PageIndicator(0, 2);
+        Utility::DirtyValue<bool> bleState {};
+        Utility::DirtyValue<bool> bleRadioEnabled {};
```
Unused?

tituscmd replied:
Yep, not anymore

mark9064 reviewed (commented):
Pretty much there I think :)

tituscmd committed:
remove album as member, remove unused members

tituscmd committed:
remove now unused music disc files

tituscmd committed:
de-clutter the SetDisconnectedUI and SetConnectedUI functions by using an array for the buttons and labels

tituscmd commented:
I made the button labels a bit more gray when the buttons are disabled, so it’s easier to tell at a glance what’s inactive.
I also cleaned up SetConnectedUI and SetDisconnectedUI by using arrays for the buttons and labels, since repeating the same line over and over was getting annoying.

I updated the first message of this PR as well so it matches the current UI, making it clearer for anyone checking it out for the first time - thanks for pointing that out @mark9064 :D

mark9064 commented:
[out.patch](https://github.com/user-attachments/files/24367919/out.patch)

Had a poke around, few suggestions here (mainly to avoid storing copies of all the buttons/labels again). I think the only substantial code change required is to fix the the progress bar narrowing issue, `lv_bar_set_range` takes int16s and currently the code assumes it can handle int32s

You can apply the changes locally with `git apply <path/to/patchfile>`

tituscmd committed:
apply patch from github, thanks mark!

tituscmd committed:
formatting

tituscmd committed:
clamp track duration to bar range to avoid any possible overflow

tituscmd committed:
Merge branch 'main' into music_redesign_2

tituscmd committed:
Merge branch 'main' into music_redesign_2

tituscmd committed:
Merge branch 'main' into music_redesign_2
